### PR TITLE
Add TRACKED_NO_CRATE_HASH and use it for `--remap-path-prefix`

### LINKED ISSUE #66955: `remap-path-prefix` will not re-compile properly in `dev` build
<details><summary>Old post</summary>

Using `remap-path-prefix` in development builds has no effect.
I'm not sure if this is intended or not, but it's not [documented](https://doc.rust-lang.org/rustc/command-line-arguments.html#--remap-path-prefix-remap-source-names-in-output).
Would write a documentation PR if this is intended.

</details>

Changing `remap-path-prefix` in `dev` mode will not properly cause re-compilation and the debug info will not change.
In `release` mode changing the flag will always cause proper re-compilation and update the debug info accordingly.

## Meta

```
rustc 1.41.0-nightly (412f43ac5 2019-11-24)
binary: rustc
commit-hash: 412f43ac5b4ae8c3599e71c6972112e9be4758fa
commit-date: 2019-11-24
host: x86_64-pc-windows-msvc
release: 1.41.0-nightly
LLVM version: 9.0
```

### PR DESCRIPTION
**Pull request by jyn514**
In https://github.com/rust-lang/rust/pull/48162, `--remap-path-prefix` was changed to be untracked by incremental compilation so that `.rlib`s built with and without it would be bitwise identical. Since `--remap-path-prefix` is used for reproducible builds, it's important that it doesn't change the output rlib.

Unfortunately, that broke incremental recompilation because we no longer rebuilt the rlib when `--remap-path-prefix` changed. To avoid that, but still make sure the output is bitwise identical, this PR adds a new `TRACKED_NO_CRATE_HASH` mode for CLI options. This invalidates the incremental cache, without adding the option to the crate hash embedded in the rlib.

- Add `TRACKED_NO_CRATE_HASH` and cbeuw's `SUBSTRUCT` idea
- Refactor the unit tests to ensure `non_default_option` isn't the default, which found a couple bugs.
- Refactor the tests to have less boilerplate
- Remove `real_rust_source_base_dir` from `Session` so it's only hashed once

I verified locally that this fixes https://github.com/rust-lang/rust/issues/66955.

r? @Aaron1011 (feel free to reassign)

### CODE REVIEW

**Comment by Aaron1011** [comment]
@jyn514: Can you add a test to `compiler/rustc_interface/src/tests.rs` to ensure that the option stays tracked?

**Comment by jyn514** [comment]
> @jyn514: Can you add a test to `compiler/rustc_interface/src/tests.rs` to ensure that the option stays tracked?

Done, and I also tested that `non_default_option` isn't the default, which found a couple bugs.

**Comment by jyn514** [comment]
Hmm, the test in `reproducible-builds` seems wrong? Presumably the .rlib is different because it's now hashing the `--remap-path-prefix` argument. Is there a reason it tests the .rlib specifically instead of some binary compiled using it?

**Comment by jyn514** [comment]
Actually I wonder if the issue is that tracked options make it all the way to .rlibs? I would expect them to only be stored in the incremental cache, not in any outputs.

**Comment by jyn514** [comment]
Hmm, it looks like this was intentionally changed in https://github.com/rust-lang/rust/pull/48162 ? @michaelwoerister do you have suggestions for invalidating the cache when this changes without also changing the output .rlib?

**Comment by michaelwoerister** [comment]
This is on my todo list. It does not look like a trivial problem.

**Comment by michaelwoerister** [comment]
It looks like we can't make the option `[TRACKED]` in it's current form (because that would break reproducible builds) but https://github.com/rust-lang/rust/pull/83813 will probably introduce a new form of tracking that should solve this problem.

**Comment by jyn514** [comment]
This is waiting on me to implement https://github.com/rust-lang/rust/pull/83813#issuecomment-824915891.

**Comment by jyn514** [comment]
> If you pull the body of the [TRACKED] branch in hash_option!() out into a function, it should be straightforward to add a [TRACKED_NO_CRATE_HASH] branch. You could then add a include_no_crate_hash_options parameter to dep_tracking_hash(). That way we won't have to change how things work in the dep-graph -- just pass true there and false in hir::map::index_hir().

@michaelwoerister hmm, there's also a `impl dep_tracking::DepTrackingHash for Options` - should that include the hash or not? Or should I remove that impl so everyone calling this has to chose explicitly?

**Comment by jyn514** [comment]
> should I remove that impl so everyone calling this has to chose explicitly?

I tried this and it doesn't work because `sub_hashes.insert` uses a trait object:
```
error[E0277]: the trait bound `options::DebuggingOptions: DepTrackingHash` is not satisfied
   --> compiler/rustc_session/src/options.rs:65:34
    |
49  | / macro_rules! top_level_options {
50  | |     (pub struct Options { $(
51  | |         $opt:ident : $t:ty [$dep_tracking_marker:ident $($warn_val:expr, $warn_text:expr)*],
52  | |     )* } ) => (
...   |
65  | |                                  &self.$opt,
    | |                                  ^^^^^^^^^^ the trait `DepTrackingHash` is not implemented for `options::DebuggingOptions`
...   |
79  | |     );
80  | | }
    | |_- in this expansion of `top_level_options!`
...
99  | / top_level_options!(
100 | |     pub struct Options {
101 | |         // The crate config requested for the session, which may be combined
102 | |         // with additional crate configurations during the compile process.
...   |
171 | |     }
172 | | );
    | |__- in this macro invocation
    |
    = note: required for the cast to the object type `dyn DepTrackingHash`
```
Maybe I should add `include_no_crate_hash: bool` to the trait itself?

**Comment by cbeuw** [comment]
For reference, this is what I did when I tried to implement it: https://github.com/rust-lang/rust/pull/83813#issuecomment-825984049 https://github.com/rust-lang/rust/commit/39ebcebebb8476dee73135973f85cab12a5a34e6

There probably is a better way though

**Comment by jyn514** [comment]
> Adding this argument to the trait's method doesn't make sense because it'll also affect all individual types implementing it.

Hmm, why do you say that? What if you want to add a `TRACKED_NO_CRATE_HASH` option to a substruct?

**Comment by cbeuw** [comment]
Because all the types used as fields in the structs will also implement `DepTrackingHash`:
https://github.com/rust-lang/rust/blob/3249d2ae8374ae72a86f78ed3268a4d7f8355229/compiler/rustc_session/src/config.rs#L2330-L2343

But it doesn't make sense to have to decide if you want `include_no_crate_hash` when you are doing `DepTrackingHash::hash()` on something like a `bool` or `String`. The types in themselves don't have any "trackedness". It's the struct containing options (i.e. `Options`, `CodegenOptions` and `DebuggingOptions`) that need to discern between tracked and untracked _fields_.

Hence you should only need to supply `include_no_crate_hash` when hashing `Options`, `CodegenOptions`, `DebuggingOptions` or any future substructs, but not anything else.

**Comment by jyn514** [comment]
Ok, this now uses your SUBSTRUCT idea from https://github.com/rust-lang/rust/pull/83813#issuecomment-825984049.

File: compiler/rustc_session/src/options.rs [hunk 1]
```diff
@@ -20,42 +20,80 @@ use std::num::NonZeroUsize;
...
+}
+
+macro_rules! hash_all_tracked {
+    ($opt_name:ident, $opt_expr:expr, $sub_hashes:expr, [UNTRACKED]) => {{}};
+    ($opt_name:ident, $opt_expr:expr, $sub_hashes:expr, [TRACKED]) => {{ insert!($opt_name, $opt_expr, $sub_hashes) }};
+    ($opt_name:ident, $opt_expr:expr, $sub_hashes:expr, [TRACKED_NO_CRATE_HASH]) => {{ insert!($opt_name, $opt_expr, $sub_hashes) }};
+    ($opt_name:ident, $opt_expr:expr, $sub_hashes:expr, [SUBSTRUCT]) => {{}};
+}
+
+macro_rules! hash_crate_hash {
```
Comment (michaelwoerister) [inline compiler/rustc_session/src/options.rs:41]: Is it possible to unify the `hash_all_tracked!` and the `hash_crate_hash!` macros by adding `for_crate_hash` as a parameter and then use `if` in the `[TRACKED_NO_CRATE_HASH]` case? I'll leave it up to you whether you want to look into this.

**Review by michaelwoerister — COMMENTED** [review]
Thanks a lot, @jyn514! This looks good to me. Do you also want to handle `Session::real_rust_source_base_dir` in this PR (it would have to be moved to the top-level options and also marked as `[TRACKED_NO_CRATE_HASH]`) or rather not?

File: compiler/rustc_session/src/options.rs (hunk 1 above)
Comment (jyn514) [reply compiler/rustc_session/src/options.rs:41]: Sure thing, done.

**Comment by jyn514** [comment]
>  Do you also want to handle Session::real_rust_source_base_dir in this PR (it would have to be moved to the top-level options and also marked as [TRACKED_NO_CRATE_HASH]) or rather not?

What do you mean by handle? As far as I can tell, `real_rust_source_base_dir` isn't an option on the CLI, it's calculated when building the session: https://github.com/rust-lang/rust/blob/1919b3f22706fee0b2c6ac3d42316545900b7734/compiler/rustc_session/src/session.rs#L1393-L1411

**Comment by jyn514** [comment]
Actually I think I figured it out, don't mind me.

**Comment by jyn514** [comment]
Ok, I cleaned up the history and also made some comments doc-comments. I can move the first and last commit to separate PRs if you like.

**Comment by michaelwoerister** [comment]
Great, thank you! It's fine by me to keep the cleanup commits in this PR.

Now we only need to add at least one end-to-end test case for this. You can either extend this one:
https://github.com/rust-lang/rust/blob/master/src/test/incremental/commandline-args.rs

Or create a separate one that works the same way. The test case needs to make sure that the main codegen unit gets recompiled if a `--remap-path-prefix` option is added or changed. This can be done via adding `#![rustc_partition_codegened(module="<NAME OF MAIN CODEGEN UNIT>", cfg="<NAME OF PASS>")]` as a top-level attribute to the main module. The test I linked is a good example.

Once we have the test case we can merge the PR.

If you are feeling adventurous you can also try to adapt this test case in the same manner:
https://github.com/rust-lang/rust/tree/master/src/test/run-make-fulldeps/incr-add-rust-src-component

i.e. add the `rustc_partition_codegened` attribute and invoke the compiler there with `-Z query-dep-graph` and the right `--cfg`s. But we can also do that in a separate PR.

**Comment by jyn514** [comment]
@Aaron1011 merging this should fix https://github.com/rust-lang/rust/issues/84341, right (in particular, https://github.com/rust-lang/rust/pull/84233/commits/39648ea467a39afa3676d900656874947c747690)? Is there a way for me to test that?

**Comment by jyn514** [comment]
Err there seem to be quite a large number of rust-src issues open - maybe https://github.com/rust-lang/rust/issues/83540 is more relevant?

**Comment by jyn514** [comment]
> Now we only need to add at least one end-to-end test case for this. You can either extend this one:

Sure thing, done. Not 100% sure I did it right, would appreciate if you could take a look :)

> i.e. add the rustc_partition_codegened attribute and invoke the compiler there with -Z query-dep-graph and the right --cfgs. But we can also do that in a separate PR.

I would prefer to do that in a follow-up - I'm not sure how query-dep-graph works, I tried it on a standalone file and it didn't seem to change anything.

**Comment by Aaron1011** [comment]
@jyn514: You should be able to copy the integration test from https://github.com/rust-lang/rust/pull/83591/files#diff-cda5f17321ba6670ff87f23aff80bb00d0492c814d29e6ba832b1bb2a7763499. However, it requires adding a perma-unstable option `-Z force-virtual-source-base` to override `CFG_VIRTUAL_RUST_SOURCE_BASE_DIR`, so that we don't have to rebuild the entire compiler.

**Comment by michaelwoerister** [comment]
> Sure thing, done. Not 100% sure I did it right, would appreciate if you could take a look :)

Looks good to me `:)`

OK, let's get this merged and follow up with further regression tests and fixes in other PRs. Thanks for implementing the fix, @jyn514!

@bors r+

File: compiler/rustc_session/src/config.rs [hunk 2]
```diff
@@ -1980,6 +1981,34 @@ pub fn build_session_options(matches: &getopts::Matches) -> Options {
...
+    // Try to find a directory containing the Rust `src`, for more details see
+    // the doc comment on the `real_rust_source_base_dir` field.
+    let tmp_buf;
+    let sysroot = match &sysroot_opt {
+        Some(s) => s,
+        None => {
+            tmp_buf = crate::filesearch::get_or_default_sysroot();
+            &tmp_buf
+        }
+    };
```
Comment (RalfJung) [inline compiler/rustc_session/src/config.rs:1993]: Computing the sysroot here is a problem as it means when the `config` driver callback changes the sysroot, that does not properly get taken into account.

I'm trying to fix this in https://github.com/rust-lang/rust/pull/122993 but I have no idea about all this "tracked" stuff, so I don't know if what I do makes any sense.

**Pull request merged**